File: 개발/backend/springboot/src/main/java/kr/co/metanet/myapp/web/HealthController.java
package kr.co.metanet.myapp.web;

import java.util.LinkedHashMap;
import java.util.Map;
import kr.co.metanet.myapp.config.AuthMode;
import kr.co.metanet.myapp.service.ItemService;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RestController;

/** 프로브·버전 엔드포인트. */
@RestController
public class HealthController {

    private final ItemService service;

    public HealthController(ItemService service) {
        this.service = service;
    }

    /**
     * 생존 프로브. <b>의존성을 확인하지 않는다</b> — 여기서 DB 를 보면
     * DB 가 잠시 끊겼을 때 파드가 재시작되어 상황이 악화된다.
     *
     * @return 200
     */
    @GetMapping("/healthz")
    public Map<String, String> healthz() {
        return Map.of("status", "ok", "env", env());
    }

    /**
     * 준비 프로브. DB 연결을 실제로 확인한다.
     *
     * @return 200 또는 503
     */
    @GetMapping("/readyz")
    public ResponseEntity<Map<String, String>> readyz() {
        try {
            if (service.ready()) {
                return ResponseEntity.ok(Map.of("status", "ready"));
            }
        } catch (RuntimeException ex) {
            return ResponseEntity.status(HttpStatus.SERVICE_UNAVAILABLE)
                    .body(Map.of("status", "not-ready", "reason", "database"));
        }
        return ResponseEntity.status(HttpStatus.SERVICE_UNAVAILABLE)
                .body(Map.of("status", "not-ready", "reason", "database"));
    }

    /**
     * 배포 버전·환경·인증 방식. 자격 증명 값은 담지 않는다.
     *
     * @return 200
     */
    @GetMapping("/version")
    public Map<String, Object> version() {
        Map<String, Object> body = new LinkedHashMap<>();
        body.put("version", System.getenv().getOrDefault("APP_VERSION", "0.0.0"));
        body.put("env", env());
        body.put("auth", AuthMode.info(System.getenv()));
        return body;
    }

    private String env() {
        return System.getenv().getOrDefault("APP_ENV", "dev");
    }
}

File: 개발/backend/nestjs/src/health/health.controller.ts
import { Controller, Get, HttpStatus, Res } from '@nestjs/common';
import type { Response } from 'express';
import { DbService } from '../db/db.service';
import { authInfo } from '../common/authmode';

/** 프로브·버전 엔드포인트. */
@Controller()
export class HealthController {
  constructor(private readonly db: DbService) {}

  /**
   * 생존 프로브. 의존성을 확인하지 않는다 — 여기서 DB 를 보면
   * DB 가 잠시 끊겼을 때 파드가 재시작되어 상황이 악화된다.
   */
  @Get('healthz')
  healthz(): { status: string; env: string } {
    return { status: 'ok', env: process.env.APP_ENV ?? 'dev' };
  }

  /** 준비 프로브. DB 연결을 실제로 확인한다. */
  @Get('readyz')
  async readyz(@Res() res: Response): Promise<void> {
    try {
      const ok = await this.db.ping();
      if (ok) {
        res.status(HttpStatus.OK).json({ status: 'ready' });
        return;
      }
    } catch {
      // 아래 공통 처리로 떨어뜨린다
    }
    res.status(HttpStatus.SERVICE_UNAVAILABLE).json({ status: 'not-ready', reason: 'database' });
  }

  /** 배포 버전·환경·인증 방식. 자격 증명 값은 담지 않는다. */
  @Get('version')
  version(): Record<string, unknown> {
    return {
      version: process.env.APP_VERSION ?? '0.0.0',
      env: process.env.APP_ENV ?? 'dev',
      auth: authInfo(process.env),
    };
  }
}
